### PR TITLE
add complete week2

### PR DESCRIPTION
**Pull request by MarcRaeder**
Changes 
Scrabble ✅
Readability ❌ Changes requested.
Caesar 🚧
Substitution 🚧

### CODE REVIEW

File: week2/scrabble/scrabble.c [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    }
+    else if (score1 < score2)
+    {
+        printf("Player 2 wins\n");
+    }
+    else
+    {
+        printf("Tie!\n");
+    }
+
```
Comment (matthiasdanne) [inline week2/scrabble/scrabble.c]:
```suggestion
        
```

File: week2/scrabble/scrabble.c [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    for (int i = 0, len = strlen(word); i < len; i++)
+    {
+        if (isupper(word[i]))
+        {
+            score += POINTS[word[i] - 'A'];
+        }
+        else if (islower(word[i]))
+        {
+            score += POINTS[word[i] - 'a'];
+        }
```
Comment (matthiasdanne) [inline week2/scrabble/scrabble.c]: Are two if - loops with basicalyl the same logic necessary?  You could rewrite your if loop, so it works regardless wether the letter is capital or not.

**Review by matthiasdanne — CHANGES_REQUESTED** [review]
I've only reviewed scrabble so far.

- Add a valid PR description

File: week2/caesar/caesar.c [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+#include <stdio.h>
+#include <ctype.h>
+#include <string.h>
+#include <stdlib.h>
+
+int main(int argc, string argv[])
+{
+    if (argc != 2)
+    {
+        printf("Usage: ./caesar key\n");
```
Comment (DevBastiS) [inline week2/caesar/caesar.c:11]:
```suggestion
        printf("Usage: ./caesar key\n");
        
```

File: week2/caesar/caesar.c [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+        return 1;
+    }
+
+    int arg_length = strlen(argv[1]);
+
+    for (int i = 0; i < arg_length; i++)
+    {
+        if (!isdigit(argv[1][i]))
+        {
+            printf("Usage: ./caesar key");
```
Comment (DevBastiS) [inline week2/caesar/caesar.c:22]:
```suggestion
            printf("Usage: ./caesar key");
            
```

File: week2/readability/readability.c [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+}
+string printLettersWordsSentences(string text)
+{
+    printf("%s\n", text);
+    printf("%i letters\n", letters);
+    printf("%i words\n", words);
+    printf("%i sentences\n", sentences);
+
+    return 0;
+}
```
Comment (DevBastiS) [inline week2/readability/readability.c]:
```suggestion
}

```

File: week2/substitution/substitution.c [hunk 6]
```diff
@@ -0,0 +1,70 @@
+#include <cs50.h>
+#include <stdio.h>
+#include <string.h>
+#include <ctype.h>
+
+int main(int argc, string argv[])
+{
+  if (argc != 2)
+  {
+    printf("Usage: ./substitution key\n");
```
Comment (DevBastiS) [inline week2/substitution/substitution.c:10]:
```suggestion
    printf("Usage: ./substitution key\n");
    
```

File: week2/substitution/substitution.c [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+    return 1;
+  }
+
+  string key = argv[1];
+  for (int i = 0; i < strlen(key); i++)
+  {
+    if (!isalpha(key[i]))
+    {
+      printf("only use letters\n");
```
Comment (DevBastiS) [inline week2/substitution/substitution.c:20]:
```suggestion
      printf("only use letters\n");
      
```

File: week2/substitution/substitution.c [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+      return 1;
+    }
+  }
+
+  if (strlen(key) != 26)
+  {
+    printf("Key must contain 26 characters.\n");
```
Comment (DevBastiS) [inline week2/substitution/substitution.c:28]:
```suggestion
    printf("Key must contain 26 characters.\n");
    
```

File: week2/substitution/substitution.c [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+    return 1;
+  }
+
+  for (int i = 0; i < strlen(key); i++)
+  {
+    for (int j = i + 1; j < strlen(key); j++)
+    {
+      if (toupper(key[i]) == toupper(key[j]))
+      {
+        printf("use each letter only once\n");
```
Comment (DevBastiS) [inline week2/substitution/substitution.c:39]:
```suggestion
        printf("use each letter only once\n");
        
```

**Review by DevBastiS — CHANGES_REQUESTED** [review]

File: week2/substitution/substitution.c [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+    }
+  }
+
+  string plaintext = get_string("plaintext: ");
+
+  for (int i = 0; i < strlen(key); i++)
+  {
+    if (islower(key[i]))
+    {
+      key[i] = key[i] - 32;
```
Comment (MKrips) [inline week2/substitution/substitution.c]:
```suggestion
      key[i] = toupper(key[i]);
```
This would be a more readable variant

**Review by MKrips — APPROVED** [review]

File: week2/readability/readability.c [hunk 11]
```diff
@@ -0,0 +1,84 @@
+#include <cs50.h>
+#include <stdio.h>
+#include <ctype.h>
+#include <string.h>
+#include <math.h>
+
+int letters = 0;
+int words = 1;
+int sentences = 0;
```
Comment (matthiasdanne) [inline week2/readability/readability.c]:
```suggestion
```

 Initializing function variables as 'global' variables should be avoided due to possible side effects. I would expect this initialization inside the main function.

You even initialize them a second time in your main function, which makes this code obsolete.

File: week2/readability/readability.c [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+
+int count_letters(string text);
+int count_words(string text);
+int count_sentences(string text);
```
Comment (matthiasdanne) [inline week2/readability/readability.c]: See above.

File: week2/readability/readability.c [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+string printLettersWordsSentences();
+
+int main(void)
+{
+    string text = get_string("Text: ");
+    int letters = count_letters(text);
+    int words = count_words(text);
+    int sentences = count_sentences(text);
+    printLettersWordsSentences(text);
+    float L = (float)letters / (float)words * 100;
```
Comment (matthiasdanne) [inline week2/readability/readability.c]:
```suggestion
    int sentences = count_sentences(text);
    
    printLettersWordsSentences(text);
    
    float L = (float)letters / (float)words * 100;
```

Improves readability.

File: week2/readability/readability.c [hunk 14]
```diff
@@ -0,0 +1,84 @@
+#include <cs50.h>
```
Comment (matthiasdanne) [inline week2/readability/readability.c:1]: Your readability solution is good. 👍

You could improve it further by reducing the amount of times you are iterating over the whole text. Right now you're iterating three times over the whole text, once in each function. You could wrap these functions with the for loop, like this:

```
for (int i = 0; i < strlen(s); i++) {
  string text = ...

  count letters(text, i)
  count words(text, i)
  count sentences(text, i)
}
```

You're passing the current index to each function alongside with the text.

**Review by matthiasdanne — CHANGES_REQUESTED** [review]
Update the PR title.

- Scrabble ✅
- Readability ❌ Changes requested.
- Caesar 🚧
- Substitution 🚧

File: week2/readability/readability.c [hunk 15]
```diff
@@ -0,0 +1,78 @@
...
+        printf("Grade %i\n", index);
+    }
+}
+
+int count_letters(string text, int letters, int i)
+{   
+    if (isalpha(text[i]))
+    {
+        letters++;
+
```
Comment (matthiasdanne) [inline week2/readability/readability.c:53]:
```suggestion
```

File: week2/readability/readability.c [hunk 16]
```diff
@@ -0,0 +1,78 @@
...
+    return letters;
+
+}
+int count_words(string text, int words, int i)
+{   
+    if (text[i] == ' ')
+    {
+        words++;
+    }
+
```
Comment (matthiasdanne) [inline week2/readability/readability.c]:
```suggestion
```

File: week2/readability/readability.c [hunk 17]
```diff
@@ -0,0 +1,78 @@
...
+    if (text[i] == ' ')
+    {
+        words++;
+    }
+    
+
+    return words;
+
+}
+int count_sentences(string text, int sentences, int i)
```
Comment (matthiasdanne) [inline week2/readability/readability.c]:
```suggestion
}

int count_sentences(string text, int sentences, int i)
```

File: week2/readability/readability.c [hunk 18]
```diff
@@ -0,0 +1,78 @@
...
+int count_words(string text, int words, int i)
+{   
+    if (text[i] == ' ')
+    {
+        words++;
+    }
+    
+
+    return words;
+
```
Comment (matthiasdanne) [inline week2/readability/readability.c]:
```suggestion
```

**Review by matthiasdanne — CHANGES_REQUESTED** [review]
Just some newline issues, otherwise readability is good. ✅

File: week2/substitution/substitution.c [hunk 19]
```diff
@@ -0,0 +1,74 @@
...
+    {
+      int letter = plaintext[i] - 65;
+      printf("%c", key[letter]);
+    }
+    else if (islower(plaintext[i]))
+    {
+      int letter = plaintext[i] - 97;
+      printf("%c", key[letter] + 32);
+    }
+    else
```
Comment (DevBastiS) [inline week2/substitution/substitution.c:70]: Missing braces here.

**Review by DevBastiS — CHANGES_REQUESTED** [review]

**Review by DevBastiS — APPROVED** [review]

**Pull request merged**